### PR TITLE
key/rev.py added

### LINKED ISSUE #50: Update key/rev.sh
https://github.com/EasyGnuPG/pgpg/blob/d108baf68dc2eb6de3d360b4bb1e06f2017d0946/src/cmd/key/rev.sh#L30

scripts from #39 can be used

### PR DESCRIPTION
**Pull request by diveshuttam**
closes #50

### CODE REVIEW

File: src/cmd/key/rev.sh [hunk 1]
```diff
@@ -27,7 +27,7 @@ Are you sure about this?" || return 1
 
     # import the revocation certificate
     sed -i "$revcert" -e "s/^:---/---/"
-    gpg --import "$revcert"
+    call_gpg key/rev.py "$revcert"
     call_fn gpg_send_keys $GPG_KEY
```
Comment (sourabhtk37) [inline src/cmd/key/rev.sh]: There doesn't seem to be a report like "successfully revoked your key".

Comment (diveshuttam) [reply src/cmd/key/rev.sh]: It wasn't earlier, I have added `print("Successfully imported revocation certificate")` in python script as I felt it was needed.

Comment (diveshuttam) [reply src/cmd/key/rev.sh]: Should I remove it?? If not, any suggestions on the change of message? I am not sure what are the more appropriate terms.

File: src/gpg/key/rev.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    c = gpg.Context()
+    with open(revcert_path) as revcert_file:
+        c.op_import(revcert_file)
+        result = c.op_import_result()
+
+        if result is None:
+            raise Exception("Could not import revocation certificate")
+        else:
+            print("Successfully imported revocation certifiate")
+
```
Comment (dashohoxha) [inline src/gpg/key/rev.py]: The user is asking to revoke the existing key. So, this message in my opinion, does not make much sense to the user. The same for the error message.

I would suggest that the python script just return an exit code (0 when successful, anything else if there is an error). The bash script can check this exit code and print a message that is appropriate.

The messages can be like this: "Key revoked successfully." and "Key revocation failed."

Comment (dashohoxha) [reply src/gpg/key/rev.py]: However I think that the message "Key revoked successfully." is redundant. Unix/Linux commands usually don't print any message when everything is Ok.

**Comment by dashohoxha** [comment]
If the python script is very short, maybe you can execute python code directly from the bash script. For the case of revocation it could be something like this:
```bash
python3 <<EndOfPythonCode
import gpg sys
gpg.Context().op_import(open('$revcert'))
EndOfPythonCode
```

File: src/gpg/key/rev.py (hunk 2 above)
Comment (diveshuttam) [reply src/gpg/key/rev.py]: Ok returning a status code seems fine to me. gpg does the same, we only issue an error message in  case of a gpgme exception/some other error occurs.

**Comment by diveshuttam** [comment]
>If the python script is very short, maybe you can execute python code directly from the bash script

I don't think it would be nice, though it shortens the code, but having python code in one place helps. Suppose I have some common changes in python scripts like adding gpgme exceptions. Or may be some specific changes due to change in GPGME itself. Also my understanding of GPGME is changing as I proceed.  It is nice to have them in one place to correct them. Moreover they don't require much of extra work to make them a file.

**Review by sourabhtk37 — DISMISSED** [review]
LGTM.

**Comment by sourabhtk37** [comment]
I noticed that if you supply incorrect or revocation cert of another key, then it will not report any error.

**Comment by diveshuttam** [comment]
Yup currently revocation certificate path is passed through egpg which in turn generated through 1) fingerprint of the key or 2) is passed by the user. 

In 1) egpg asks for the key ID to gpg only (which will change to gpgme after rewrite). So currently I am basing this pr on the assumption that keyid is correct(because it is being fetched by gpg, which seems reasonable), later the I'll put any checks required in getting fingerprint form gpgme.
In 2) Also the path to revocation certificate is passed by egpg to the python script and exists(with high probability) as it is already being checked in the rev.sh a few lines before calling. (This point is valid to 1 also)

**Comment by diveshuttam** [comment]
@sourabhtk37 I have fixed it using the debug variable, I print the traceback only if `DEBUG='yes'`. This is because long and detailed python tracebacks are not very much useful for normal users. (considering that are using egpg because it is easier). It just prints "Key revocation failed" as of now. Later we may add specific messages like "Internal Error" or "Revocation Certificate not found" etc. based on the exception that is raised.
If one wants to see the traceback and details s/he can use 'egpg set debug yes'.

Currently I am using BaseException as I don't know what exceptions can be raised by gpgme in this scenario. Will narrow this down as we test this. Also could not test any valid exceptions, FileNotFound is valid but unlikely(for revocation certificate) as we are checking it in egpg code only a few lines earlier. (Refer @dashohoxha 's comment on #57)

File: src/gpg/key/rev.py [hunk 3]
```diff
@@ -1,13 +1,19 @@
...
-        if result is None:
-            exit(1)
+    try:
+        with open(revcert_path) as revcert_file:
+            c.op_import(revcert_file)
+            result = c.op_import_result()
+            if result is None:
+                exit(1)
+    except BaseException:
+        if(os.environ['EGPG_DEBUG']=='yes'):
```
Comment (dashohoxha) [inline src/gpg/key/rev.py]: Why not use `DEBUG`?

File: src/egpg.sh [hunk 4]
```diff
@@ -192,6 +192,7 @@ config() {
 
     export GNUPGHOME
     export GPG_TTY=$(tty)
+    export EGPG_DEBUG=$DEBUG
```
Comment (dashohoxha) [inline src/egpg.sh]: Why introduce a new variable?

File: src/cmd/key/rev.sh [hunk 5]
```diff
@@ -27,7 +27,11 @@ Are you sure about this?" || return 1
 
     # import the revocation certificate
     sed -i "$revcert" -e "s/^:---/---/"
-    gpg --import "$revcert"
+    call_gpg key/rev.py "$revcert"
+    
+    local err=$?
+    [[ $err == 0 ]] || fail "Key revokation failed"
```
Comment (dashohoxha) [inline src/cmd/key/rev.sh]: It should be `revocation`.

File: src/egpg.sh (hunk 4 above)
Comment (diveshuttam) [reply src/egpg.sh]: I want to test for DEBUG in all python scripts. $DEBUG is not already exported. So did it this way as variable name DEBUG is very common variable used by other programs also, I didn't wanted it to interfere with the already running programs.

UPDATE: I was assuming that these variables get reflected in the main shell. But I was wrong I tried this and saw that calling egpg launches its own bash instance. So $DEBUG will not interfere with the programs from the main shell, I can export DEBUG also.

Comment (diveshuttam) [reply src/egpg.sh]: I had kind of misinterpreted working of calls in bash, I knew that environments are only propagated downwards but I forgot to consider that each call in Bash launches a new subprocess and hence exporting a variable won't affect the main shell. Have corrected this and now  we are using variable $DEBUG only.

**Comment by diveshuttam** [comment]
Are the changes ok??

**Review by dashohoxha — APPROVED** [review]
Sorry. I though this was approved (either by me or by T K Sourabh). Actually when new commits are added approval expires and a new approval is needed.

These changes seem Ok, you can merge them.

**Pull request merged**